UnifiedPDF: Implement link following (external links)

hortont424 opened the pull request:
#### 869c3738830502f32e8eb2b2e2fc7ae1afb773c4
<pre>
UnifiedPDF: Implement link following (external links)
<a href="https://bugs.webkit.org/show_bug.cgi?id=268894">https://bugs.webkit.org/show_bug.cgi?id=268894</a>
<a href="https://rdar.apple.com/118550787">rdar://118550787</a>

Reviewed by Richard Robinson.

* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.h:
Hoist m_lastMouseEvent to PDFPluginBase.

* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
(WebKit::PDFPlugin::clickedLink): Deleted.
Hoist clickedLink up to PDFPluginBase.

(-[WKPDFLayerControllerDelegate pdfLayerController:clickedLinkWithURL:]):
Adopt it in WKPDFLayerControllerDelegate.

(WebKit::PDFPlugin::nsEventForWebMouseEvent):
(WebKit::PDFPlugin::handleMouseEvent):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::lastKnownMousePositionInView const):
Get rid of m_lastMousePositionInPluginCoordinates; since we have to store the
whole event, we&apos;ll just use that instead.

(WebKit::PDFPluginBase::navigateToURL): Hoist!

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
(WebKit::AnnotationTrackingState::trackedAnnotation const):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::handleMouseEvent):
Update the cursor to the pointer when we stop tracking an annotation.
We&apos;ll start updating again after the mouseup.

(WebKit::UnifiedPDFPlugin::didClickLinkAnnotation):
Follow the link! Later we&apos;ll follow internal links too.

(WebKit::AnnotationTrackingState::finishAnnotationTracking):
Remove this assertion so we can track other types of annotations that
don&apos;t do anything here.

Canonical link: <a href="https://commits.webkit.org/274205@main">https://commits.webkit.org/274205@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5d433fbb8c5650667494fccd82c8316275e1e59c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/38268 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/17212 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/40584 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/40826 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/34042 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/19969 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/14552 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/32289 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/38841 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/14531 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 api-mac](https://ews-build.webkit.org/#/builders/API-Tests-macOS-EWS "Waiting to run tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/12645 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/12630 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/42105 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/34816 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/34639 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/38464 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/13208 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/10920 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/36661 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/14762 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8610 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/13621 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/14227 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

sammygill commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1065,10 +1074,15 @@ static AffineTransform documentSpaceToPageSpaceTransform(const IntDegrees& pageR
         switch (auto mouseEventButton = event.button()) {
         case WebMouseEventButton::Left:
-            if (m_annotationTrackingState.trackedAnnotation()) {
+            if (auto trackedAnnotation = m_annotationTrackingState.trackedAnnotation()) {
```
We might want to use `RetainPtr` here or swap the order of operations below since `finishAnnotationTracking` clears the annotation it is holding onto

sammygill replied:
or maybe change the API of `trackedAnnotation()`

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -871,6 +872,30 @@ static void verboseLog(PDFIncrementalLoader* incrementalLoader, uint64_t streame
+    if (url.protocolIsJavaScript())
+        return;
+
+    auto* frame = m_frame ? m_frame->coreLocalFrame() : nullptr;
```
I know you didn't write this, but
```suggestion
    RefPtr frame = m_frame ? m_frame->coreLocalFrame() : nullptr;
```

aprotyas reviewed (commented):
LGTM

hortont424 committed:
UnifiedPDF: Implement link following (external links)
https://bugs.webkit.org/show_bug.cgi?id=268894
rdar://118550787

Reviewed by Richard Robinson.

* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.h:
Hoist m_lastMouseEvent to PDFPluginBase.

* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
(WebKit::PDFPlugin::clickedLink): Deleted.
Hoist clickedLink up to PDFPluginBase.

(-[WKPDFLayerControllerDelegate pdfLayerController:clickedLinkWithURL:]):
Adopt it in WKPDFLayerControllerDelegate.

(WebKit::PDFPlugin::nsEventForWebMouseEvent):
(WebKit::PDFPlugin::handleMouseEvent):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::lastKnownMousePositionInView const):
Get rid of m_lastMousePositionInPluginCoordinates; since we have to store the
whole event, we'll just use that instead.

(WebKit::PDFPluginBase::navigateToURL): Hoist!

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
(WebKit::AnnotationTrackingState::trackedAnnotation const):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::handleMouseEvent):
Update the cursor to the pointer when we stop tracking an annotation.
We'll start updating again after the mouseup.

(WebKit::UnifiedPDFPlugin::didClickLinkAnnotation):
Follow the link! Later we'll follow internal links too.

(WebKit::AnnotationTrackingState::finishAnnotationTracking):
Remove this assertion so we can track other types of annotations that
don't do anything here.

Canonical link: https://commits.webkit.org/274205@main
